### PR TITLE
Delete temporary packfile in indexer

### PR DESCRIPTION
**Pull request by lhchavez**
This change deletes the temporary packfile that the indexer creates to
avoid littering the pack/ directory with garbage.

### CODE REVIEW

**Comment by carlosmn** [comment]
What errors are you seeing that cause this function to fail? It should be exceedingly rare that anything in this function fails and the most likely culprit would be the creation of this file itself.

**Comment by lhchavez** [comment]
in `git_indexer_new`? nothing I can think of, it's just there for correctness, but I can remove it if needed. `git_indexer_free` will always leak the file if it `git_indexer_commit` was not called before (say, because you were writing a validator for a pack that was uploaded, and although the packfile was correct, the commit itself would fail validation).

**Comment by lhchavez** [comment]
Now with a test that actually fails if I remove the check in `git_indexer_commit`. PTAL!

**Comment by lhchavez** [comment]
now the tests actually pass -_-. PTAL.

File: src/indexer.c [hunk 1]
```diff
@@ -151,6 +151,12 @@ int git_indexer_new(
 	if (fd != -1)
 		p_close(fd);
 
+	if (git_buf_is_allocated(&tmp_path))
```
Comment (ethomson) [inline src/indexer.c]: `git_buf_len(&tmp_path)` is correct here.

`is_allocated` literally looks at whether we have `malloc`ed the structure, not whether it has contents.  These are orthogonal concepts:  the buffer may point to static memory and thus be nonempty but not allocated, similarly, we could allocate a buffer, set contents into it, then clear the contents, but it would still be allocated.

File: src/indexer.c [hunk 2]
```diff
@@ -151,6 +151,12 @@ int git_indexer_new(
...
+		(void)p_unlink(git_buf_cstr(&tmp_path));
```
Comment (ethomson) [inline src/indexer.c]: We do not cast unused return values to `void`.  There are a few other places in here that should be removed also.

File: tests/pack/indexer.c [hunk 3]
```diff
@@ -125,3 +125,51 @@ void test_pack_indexer__fix_thin(void)
...
+{
+	git_indexer *idx = NULL;
+	git_buf path = GIT_BUF_INIT;
+	git_buf first_tmp_file = GIT_BUF_INIT;
+
+	/* Precondition: there are no temporary files. */
+	cl_git_pass(git_buf_sets(&path, clar_sandbox_path()));
+	cl_git_pass(find_tmp_file_recurs(&first_tmp_file, &path));
+	git_buf_free(&path);
+	if (git_buf_is_allocated(&first_tmp_file)) {
```
Comment (ethomson) [inline tests/pack/indexer.c]: I wouldn't `cl_skip` here.  Do you expect temp files might exist?  Then please don't skip the test, fix it.  If not, assert that they don't so that we fail.

File: tests/pack/indexer.c [hunk 4]
```diff
@@ -125,3 +125,51 @@ void test_pack_indexer__fix_thin(void)
...
+	git_indexer_free(idx);
+
+	cl_git_pass(git_buf_sets(&path, clar_sandbox_path()));
+	cl_git_pass(find_tmp_file_recurs(&first_tmp_file, &path));
+	git_buf_free(&path);
+	if (git_buf_is_allocated(&first_tmp_file)) {
+		cl_warning(git_buf_cstr(&first_tmp_file));
+		git_buf_free(&first_tmp_file);
+		cl_fail("Found a temporary file");
+	}
```
Comment (ethomson) [inline tests/pack/indexer.c]: There's no need to free things on failures, we don't bother testing for memory leaks if it doesn't even work.  This 5 lines can be more easily rewritten as:

```
cl_assert(git_buf_len(&first_tmp_file) > 0)
git_buf_free(&first_tmp_file);
```

(Which, I think, is probably what the test above should look like.)

File: tests/pack/indexer.c (hunk 3 above)
Comment (lhchavez) [inline tests/pack/indexer.c]: i expect they shouldn't exist (and currently don't), but cannot make that guarantee currently (I opened https://github.com/libgit2/libgit2/issues/4066 for that).

If we are cool with this test failing when other tests leak files, i can change the skip to a warning to make it clear that it wasn't this test that caused the failure, but another one.

File: src/indexer.c (hunk 1 above)
Comment (lhchavez) [inline src/indexer.c]: Done

File: src/indexer.c (hunk 2 above)
Comment (lhchavez) [inline src/indexer.c]: Done (saw this was used inconsistently in a few other places).

File: tests/pack/indexer.c (hunk 3 above)
Comment (lhchavez) [inline tests/pack/indexer.c]: After a second thought, if this fails, it is obvious from the comment in L154 that it's not this test's fault, so changed it to simple assert with no log after all.

File: tests/pack/indexer.c (hunk 4 above)
Comment (lhchavez) [inline tests/pack/indexer.c]: Done

**Comment by ethomson** [comment]
I added a quick check before doing the `unlink` to avoid calling it when we have already moved the file into place.  I merged this manually; thanks again for fixing this!
